[chip dv] Auto-generated interrupt test

sriyerg opened the pull request:
This is the starting point of the SW interrupt test that will be
auto-generated. Currently, it only caters to UART0. Submitting draft PR
for early feedback, before I add it as a template to the topgen script.

EDIT: In the latest force push, the commit has been updated such that the 
test is now autogenerated using the topgen.py utility. 

Signed-off-by: Srikrishna Iyer <sriyer@google.com>

sriyerg commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,183 @@
+static volatile dif_uart_irq_t uart_irq_expected;
+static volatile dif_uart_irq_t uart_irq_serviced;
+
+/**
+ * UART irq handler.
+ */
+static void uart_isr(dif_uart_t *uart, dif_rv_plic_irq_id_t plic_irq_id) {
+  // Correlate the interrupt fired at PLIC with UART.
+  dif_uart_irq_t uart_irq =
+      (dif_uart_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                         kTopEarlgreyPlicIrqIdUart0TxWatermark);
+  CHECK(uart_irq == uart_irq_expected,
+        "Incorrect UART irq received at PLIC: exp = %d, obs = %d",
+        uart_irq_expected, uart_irq);
+  uart_irq_serviced = uart_irq;
+
+  // Check if the same interrupt fired at UART as well.
+  bool is_pending;
+  CHECK(dif_uart_irq_is_pending(uart, uart_irq, &is_pending) == kDifUartOk,
+        "dif_uart_irq_is_pending failed");
+  CHECK(is_pending, "UART interrupt fired at PLIC did not fire at UART");
+
+  // TODO: read the whole status register to confirm none other pending.
+
+  // Clear the interrupt at UART.
+  CHECK(dif_uart_irq_acknowledge(uart, uart_irq) == kDifUartOk,
+        "dif_uart_irq_acknowledge failed");
+}
```
@timothytrippel, do you think we could autogen this function for all blocks in the testutils area and a common ISR API that can be used in tests directly?

timothytrippel replied:
Yes absolutely. Created task in #8142 to track this. I will create a separate PR for this.

sriyerg commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,183 @@
+            },
+            &uart0) == kDifUartOk,
+        "dif_uart_init failed");
+  for (dif_uart_irq_t uart_irq = kDifUartIrqTxWatermark;
+       uart_irq <= kDifUartIrqRxParityErr; ++uart_irq) {
+    CHECK(dif_uart_irq_set_enabled(&uart0, uart_irq, kDifUartToggleEnabled) ==
+              kDifUartOk,
+          "dif_uart_irq_set_enabled failed");
+  }
```
@timothytrippel, it will be nice to have one function that just enables all interrupts as a DIF.

timothytrippel replied:
As you suggested, just use dif_*_irq_restore_all(<all-1s>).

weicaiyang reviewed (commented):
This looks great to me. Defer to @timothytrippel for approval.

cindychip reviewed (commented):
Thanks for the implementation. LGTM. Let us see if SW side has any comments.

timothytrippel commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,183 @@
+static void plic_irqs_enable(void) {
+  mmio_region_t plic_base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR);
+  CHECK(dif_rv_plic_init((dif_rv_plic_params_t){.base_addr = plic_base_addr},
```
Just a heads up, as a part of the IRQ DIF generation/refactoring an IP's base address will be located directly in the dif_<ip>_t struct, instead of being nested in the dif_<ip>_params_t struct. Also shared result types will be used across most diff functions What this means it that the function prototype:

dif_uart_result_t dif_uart_init(dif_uart_params_t params, dif_uart_t *uart)

will become:

dif_result_t dif_uart_init(mmio_region_t base_addr, dif_uart_t *uart) .

Just an FYI. These changes specifically for the UART will land in #8193. Feel free to wait to merge this on top of #8193, or I can refactor this as a part #8193 if this is merged first. Either works for me!

sriyerg replied:
Ack, thanks. I will wait for your PRs to merge, then I will rebase on top.

timothytrippel commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,183 @@
+  // UART0.
+  mmio_region_t uart_base_addr =
+      mmio_region_from_addr(TOP_EARLGREY_UART0_BASE_ADDR);
+  CHECK(dif_uart_init(
```
same here

timothytrippel commented:
A couple comments above, otherwise LGTM. Thanks for doing this!

timothytrippel commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,211 @@
+ * Declared volatile because it is referenced in the main program flow as well
+ * as the ISR.
+ */
+static volatile uint32_t irq_expected;
+static volatile uint32_t irq_serviced;
```
Is there a reason these can't be dif_##_peripheral##_irq_t types throughout? (to avoid casting)

sriyerg replied:
It can be, but it will end up getting expanded to 70 variables (2 per IP). That's probably ok though. Updated.

timothytrippel commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,211 @@
+/**
+ * Triggers all irqs in all peripherals one by one.
+ *
+ * Walks through all instanances of all peripherals and triggers an interupt one
```
(nit) instanances --> instances
(nit) interupt --> interrupt

timothytrippel reviewed (approved):
couple of minor nits, otherwise LGTM!

sriyerg commented:
Updated the test to be fully auto-generated with `topgen` in the latest force-push.

timothytrippel commented on sw/device/tests/autogen/meson.build:
```diff
@@ -0,0 +1,48 @@
+    'plic_all_irqs_test_lib',
+    sources: ['plic_all_irqs_test.c'],
+    dependencies: [
+      # sw_lib_dif_adc_ctrl,
+      # sw_lib_dif_alert_handler,
+      # sw_lib_dif_aon_timer,
+      # sw_lib_dif_csrng,
+      # sw_lib_dif_edn,
+      # sw_lib_dif_entropy_src,
+      # sw_lib_dif_flash_ctrl,
+      # sw_lib_dif_gpio,
+      # sw_lib_dif_hmac,
+      # sw_lib_dif_i2c,
+      # sw_lib_dif_keymgr,
+      # sw_lib_dif_kmac,
+      # sw_lib_dif_otbn,
+      # sw_lib_dif_otp_ctrl,
+      # sw_lib_dif_pattgen,
+      # sw_lib_dif_pwrmgr,
+      # sw_lib_dif_rv_timer,
+      # sw_lib_dif_spi_device,
+      # sw_lib_dif_spi_host,
+      # sw_lib_dif_sysrst_ctrl,
+      sw_lib_dif_uart,
+      # sw_lib_dif_usbdev,
```
nice! good to prep this.

timothytrippel commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,489 @@
+#define PERIPHERAL_INIT(_peripheral, _handle, _base_addr)          \
+  do {                                                             \
+    mmio_region_t base_addr = mmio_region_from_addr(_base_addr);   \
+    CHECK(dif_##_peripheral##_init(base_addr, &_handle) == kDifOk, \
```
Just a heads up/FYI, this should be the DIF init function prototype for _most_ IPs, however for IPs that are templated, thus have HW instantiation params (*_params_t), there will be an extra parameter in this function prototype. See #8286 for an example with the alert_handler. Just means that this will have some special cases thats all... square peg, round hole...

timothytrippel commented:
Updates LGTM, thanks!

moidx commented:
@mcy can you take a look at this proposal to automate the generation of [chip_plic_all_irqs](https://cs.opensource.google/opentitan/opentitan/+/master:hw/top_earlgrey/data/chip_testplan.hjson?q=chip_plic_all_irqs) chip level tests?

sriyerg commented:
Added `clang_format()` function to `topgen` in the latest force-push so that the generated C test is formatted and the CI does not complain. Also added warning header in the test indicating how was this test generated. 

Looks like we disable clang-formatting on other auto-generated sources such as top level header. Was there a specific reason to disable it? @cfrantz @moidx @mcy

mcy commented on sw/device/tests/autogen/meson.build:
```diff
@@ -0,0 +1,48 @@
+    'plic_all_irqs_test_lib',
+    sources: ['plic_all_irqs_test.c'],
+    dependencies: [
+      # sw_lib_dif_adc_ctrl,
```
Is the expectation that someone will be manually switching these as they become necessary, or that a script will regenerate this file? Seems like the latter. Probably worth marking all generated files as such so people don't get confused.

mcy replied:
Kinda confused about what the comments are achieving if this isn't human editable. Humans should never look at these files, so why put comments in them?

mcy replied:
Right but... the file will get regenerated, right? Like I think we should not be emitting these comments at all.

sriyerg replied:
Don't really care either way. Just want this PR merged today, so, whatever it takes. Removed.

mcy commented on util/topgen/c_test.py:
```diff
@@ -0,0 +1,94 @@
+from .lib import Name
+
+
+class TestPeripheral:
+    """Captures a peripheral instance's attributes for use in SW tests.
+
+    It encapsulates the settings for each peripheral's instance in the design.
+    These settings are captured as identifiers that are already defined in
+    other autogenerated / hand-edited sources such as top_{name}.h and the DIF
+    headers, rather than hard-coded constants. This is done to ensure that a
+    single source of truth is referenced in all of the tests.
+    """
+    def __init__(self, name: str, inst_name: str, base_addr_name: str):
+        self.name = name
+        self.inst_name = inst_name
+        self.base_addr_name = base_addr_name
+
+
+class IrqTestPeripheral(TestPeripheral):
```
Is there a reason this is split into two classes? From reading this code, it's very hard to understand why it's organized in this manner. These classes just look like configuration data to me?

mcy commented on util/topgen/templates/plic_all_irqs_test.c.tpl:
```diff
@@ -0,0 +1,246 @@
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
+        peripheral_expected, peripheral);
+<%text>
+#define PERIPHERAL_ISR(_peripheral, _handle, _plic_start_irq)                 \
```
So, the question to ask here: why is this jammed into a generated file, instead of being in a "real" header? It doesn't seem to depend on anything in the template, so maybe it makes sense to inline it into where it's used below, or put it in its own header.

Whatever can be hoisted out of this header probably should.

Also, is there a reason for the leading underscore names? We don't have specfic style guidance here, so we should be consistent with the rest of the project.

timothytrippel replied:
After thinking about this more from our discussion earlier, I think the style guides are confusing here. Specifically, defining these macros right before they are used, and undefining them immediately after is directly inline with the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros), for which we mostly follow. However, there are a couple sentences in our [(OpenTitan) style guide](​​https://docs.opentitan.org/doc/rm/c_cpp_coding_style/) that seem to toss this guidance out the window, and enable, as @mcy suggests, hoisting the macros to a public header: "_In contrast to the recommendation in the Google C++ style guide, exporting macros as part of the public API is allowed in C code. A typical use case is a header with register definitions._", though the exact use case here is different.

Regardless of my style guide naivete, since this is autogenerated code, I think it makes sense to keep this inlined in the Mako template for loop context you make use of below (as @mcy suggests), and just do away with the macro entirely. This would naturally also remove the leading underscore names :) Likewise with the other macro blocks.

mcy replied:
> I did not see any real benefit to moving these to a separate header since they are not going to be used in any other tests, hence no need to make these public.

There is a benefit: they are now real C files with real indexing. The templates are very hard to read, and once we move to Bazel we will probably stop checking generated code in in a lot of cases. Please factor the macros out into a header.

> The leading underscore helps distinguish macro args from the identifiers in the rest of the test code.

I know, but the rest of the project doesn't do this. If you feel strongly, send me a PR to the style guide and fixes to make the entire project consistent with this style. I can see the benefit, but OT is small enough that consistency is important.

sriyerg replied:
Ok. Moved the macros to headers, and removed leading underscores.

mcy commented on util/topgen/templates/plic_all_irqs_test.c.tpl:
```diff
@@ -0,0 +1,246 @@
+</%text>\
+
+  switch (peripheral) {
+% for p in helper.irq_peripherals:
+<%
+  comment = '' if p.name in enabled_peripherals else '// '
+%>\
+    ${comment}case ${p.plic_name}:
+    ${comment}  PERIPHERAL_ISR(${p.name}, ${p.inst_name}, ${p.plic_start_irq});
+    ${comment}  break;
+% endfor
```
Throughout: we may want to match the indentation with all the Mako directives and the C in this file.

mcy commented on util/topgen/templates/plic_all_irqs_test.c.tpl:
```diff
@@ -0,0 +1,246 @@
+/**
+ * Enables all IRQs at PLIC.
+ */
+static void plic_irqs_enable(void) {
```
Functions like this one should probably be hoisted into another target, just like the macros. Let's keep things that don't depend on template input separate.

mcy commented on util/topgen/templates/plic_all_irqs_test.c.tpl:
```diff
@@ -0,0 +1,246 @@
+  switch (peripheral) {
+% for p in helper.irq_peripherals:
+<%
+  comment = '' if p.name in enabled_peripherals else '// '
```
Seeing this repeated a lot. I think it may be worthwhile to define a mako `<%def>`? It'd be nice to just write `${comment(p)}` or similar.

msfschaffner commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,518 @@
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == peripheral_expected,
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
```
nit: Interurpt -> Interrupt

msfschaffner commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,518 @@
+    /* TODO: read the whole status register to confirm none other pending. */ \
+    bool is_pending;                                                          \
+    CHECK(dif_##_peripheral##_irq_is_pending(&_handle, irq, &is_pending) ==   \
+              kDifOk,                                                         \
```
nit: indentation seems to be off here

msfschaffner commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,518 @@
+/**
+ * Triggers all IRQs in all peripherals one by one.
+ *
+ * Walks through all instannces of all peripherals and triggers an interrupt one
```
nit: instannces -> instances

msfschaffner commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,518 @@
+    /* Clear the interrupt at the peripheral. */                              \
+    CHECK(dif_##_peripheral##_irq_acknowledge(&_handle, irq) == kDifOk,       \
+          "dif_" #_peripheral "_irq_acknowledge failed");                     \
+  } while (0)
```
is there a particular reason this is enclosed in a `do { ... } while (0)` loop? or could we just use a `{ ... }` block instead? (sorry I may be missing something...)

timothytrippel replied:
I think this is common technique to avoid some gotchas, see here: http://www.bruceblinn.com/linuxinfo/DoWhile.html

msfschaffner commented on sw/device/tests/autogen/plic_all_irqs_test.c:
```diff
@@ -0,0 +1,518 @@
+  // kTopEarlgreyPlicPeripheralSysrstCtrlAon,
+  //                         kDifSysrstCtrlIrqSysrstCtrl,
+  //                         kDifSysrstCtrlIrqSysrstCtrl);
+  PERIPHERAL_IRQS_TRIGGER(uart, uart0, kTopEarlgreyPlicPeripheralUart0,
```
does it make sense to call these triggers in pseudo-random order at some point?

msfschaffner replied:
right, if we mainly want to test connectivity here, this should be good as is.

msfschaffner commented on util/topgen/c_test.py:
```diff
@@ -0,0 +1,94 @@
+            # Device regions may have multiple TL interfaces. Pick the region
+            # associated with the 'core' interface.
+            if_name = None if len(
+                self.device_regions[inst_name]) == 1 else 'core'
```
this choice is not standardized. maybe throw an error or warning if more than 2 TL-UL interfaces are present, and none of them is called `core`?

msfschaffner replied:
ok good, thanks for adding.

msfschaffner reviewed (commented):
Thanks for adding this, @sriyerg. I have a couple of small comments / questions.

sriyerg commented:
In the force push 30 minutes ago, I addressed all the pending comments. 
In the latest force push, I added the test to Verilator and FPGA targets.

mcy commented on util/topgen.py:
```diff
@@ -37,6 +39,8 @@
 from topgen.resets import Resets
 
 # Common header for generated files
+# TODO; warnhdr below is 100 chars which exceeds the 80 char limit used by
+# clang-format. Refactor this in a separate PR.
```
Todos should be written in more of an "infinitive" tense, and include an associated bug:
```
# TODO(###)
```

"Refactor this in a separate PR" is redundant.

mcy commented on util/topgen/templates/meson.build.tpl:
```diff
@@ -0,0 +1,39 @@
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+${gencmd.replace("//", "#")}
+<%text>
```
Why do you need a %text escaping directive here?

mcy replied:
But not for the license? I think there's something wrong here. Is it the multiple ##s?

sriyerg replied:
Fixed.

mcy commented on util/topgen/templates/meson.build.tpl:
```diff
@@ -0,0 +1,39 @@
+    'plic_all_irqs_test_lib',
+    sources: ['plic_all_irqs_test.c'],
+    dependencies: [
+<%
+irq_peripheral_names = sorted({p.name for p in helper.irq_peripherals})
+enabled_peripherals = ['uart']
+%>\
+% for p in irq_peripheral_names:
+<%
+  comment = '' if p in enabled_peripherals else '# '
+%>\
+      ${comment}sw_lib_dif_${p},
+% endfor
```
Is it worth indenting all of the directives to make this easier to read?

sriyerg commented on sw/device/tests/autogen/meson.build:
```diff
@@ -0,0 +1,54 @@
+    'plic_all_irqs_test_lib',
+    sources: ['plic_all_irqs_test.c'],
+    dependencies: [
+                  # sw_lib_dif_adc_ctrl,
```
Indenting Mako directives with inline text makes the indentation in the generated text look horrible. It works for C because `clang-format` but not for other sources. I think we should stick to left-aligned mako code in templates.

mcy commented on sw/device/tests/plic_all_irqs_test.h:
```diff
@@ -0,0 +1,85 @@
+#define OPENTITAN_SW_DEVICE_TESTS_PLIC_ALL_IRQS_TEST_H_
+
+/**
+ * Helper macros for the auto-generated plic_all_irqs_test. These are not meant
```
Maybe the file should have a name that reflects that? The file name seems to imply that it's an API, rather than being a helper. (E.g., `plic_all_irqs_test_helpers.h`.)

sriyerg replied:
Done.

mcy commented on sw/device/tests/plic_all_irqs_test.h:
```diff
@@ -0,0 +1,85 @@
+      /* NOTE; The interrupt triggered above is instantaneous - Ibex jumps     \
+       * into the ISR even before a WFI can be called. Hence, we add a couple  \
+       * of nops instead. */                                                   \
+      asm volatile("nop");                                                     \
+      asm volatile("nop");                                                     \
```
Hang on, why do you need these no-ops? Presumably Ibex enters the handler before irq_force even returns, so it's not clear what effect these are supposed to have.

sriyerg replied:
Ahh yes, I was experimenting with these. They are not needed. Removed.

timothytrippel commented on sw/device/tests/plic_all_irqs_test_helper.h:
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_TESTS_PLIC_ALL_IRQS_TEST_HELPER_H_
+#define OPENTITAN_SW_DEVICE_TESTS_PLIC_ALL_IRQS_TEST_HELPER_H_
```
maybe this should be in sw/device/lib/testing/rv_plic_all_irqs_testutils.h ?

sriyerg replied:
If I did, it would imply that other tests can use these macros, which is not the case. Perhaps leave it here?

msfschaffner reviewed (approved):
LGTM from my side

sriyerg commented on util/topgen/templates/plic_all_irqs_test.c.tpl:
```diff
@@ -0,0 +1,154 @@
+    .escalation_signal_count = ALERT_HANDLER_PARAM_N_ESC_SEV};
+
+  % for p in helper.irq_peripherals:
+  % if p.is_templated:
```
Assumption here that all templated IPs are parameterized w.r.t. DIFs.
